commit f13b2b61da12bc4368908b9eae19754ac44c291f
Author: Nicolas CHARLES <nicolas.charles@normation.com>
Date:   Tue Nov 4 11:10:46 2014 +0100

    Backporting commit 37065c7ac536a8147179eeb091d985abe62d887e from Volker Hilsheimer into CFEngine 3.6.0

diff --git a/cf-agent/files_editline.c b/cf-agent/files_editline.c
index 5ba21e2..6daa826 100644
--- a/cf-agent/files_editline.c
+++ b/cf-agent/files_editline.c
@@ -1650,9 +1650,11 @@ static int InsertCompoundLineAtLocation(EvalContext *ctx, char *chunk, Item **st
     bool retval = false;
     char buf[CF_EXPANDSIZE];
     char *sp;
-    int preserve_block = a.sourcetype && (strcmp(a.sourcetype, "preserve_all_lines") == 0 || strcmp(a.sourcetype, "preserve_block") == 0 || strcmp(a.sourcetype, "file_preserve_block") == 0);
 
-    if (!preserve_block && MatchRegion(ctx, chunk, location, NULL, false))
+    int preserve_all_lines = a.sourcetype && strcmp(a.sourcetype, "preserve_all_lines") == 0;
+    int preserve_block = a.sourcetype && (preserve_all_lines || strcmp(a.sourcetype, "preserve_block") == 0 || strcmp(a.sourcetype, "file_preserve_block") == 0);
+
+    if (!preserve_all_lines && MatchRegion(ctx, chunk, location, NULL, false))
     {
         cfPS(ctx, LOG_LEVEL_VERBOSE, PROMISE_RESULT_NOOP, pp, a, "Promised chunk '%s' exists within selected region of %s (promise kept)", pp->promiser, edcontext->filename);
         return false;
